secp256kfun/src/poly.rs

```diff
@@ -45,25 +46,92 @@ pub mod scalar {
+    /// secret share is associated with a participant index (index, share).
     pub fn interpolate_and_eval_poly_at_0(
-        secrets_at_indices: Vec<(Scalar<Public>, Scalar<Secret, impl ZeroChoice>)>,
+        x_and_y: &[(Scalar<Public>, Scalar<impl Secrecy, impl ZeroChoice>)],
     ) -> Scalar<Secret, Zero> {
-        let indicies: Vec<_> = secrets_at_indices.iter().map(|(index, _)| *index).collect();
-        secrets_at_indices
-            .into_iter()
+        let indicies = x_and_y.iter().map(|(index, _)| *index);
+        x_and_y
+            .iter()
             .map(|(index, secret)| {
-                let lambda = eval_basis_poly_at_0(index, indicies.iter());
+                let lambda = eval_basis_poly_at_0(*index, indicies.clone());
                 s!(secret * lambda)
             })
-            .fold(s!(0), |acc, contribution| s!(acc + contribution))
+            .fold(s!(0), |interpolated_poly, scaled_basis_poly| {
+                s!(interpolated_poly + scaled_basis_poly)
+            })
+    }
+
+    /// Interpolate a polynomial that runs through a list of `(x, y)` coordinate points represented
+    /// as scalars. In most protocols the `x` is usually known to everyone in the protocol while the
+    /// `y` is the party's signing key.
+    pub fn interpolate<S: Secrecy>(
+        x_and_y: &[(Scalar<Public>, Scalar<S, impl ZeroChoice>)],
+    ) -> Vec<Scalar<S, Zero>> {
+        let x_ms = x_and_y.iter().map(|(index, _)| *index);
+        let mut interpolating_polynomial: Vec<Scalar<S, Zero>> = vec![];
+        for (x_j, y_j) in x_and_y {
+            let basis_poly = super::lagrange_basis_poly(*x_j, x_ms.clone());
+            let scaled_basis_poly = basis_poly
+                .iter()
+                .map(|coeff| s!(coeff * y_j))
+                .collect::<Vec<_>>();
+            self::add_in_place(&mut interpolating_polynomial, scaled_basis_poly);
+        }
+
+        interpolating_polynomial
+    }
+
+    /// Multiplies two polynomials of the same [`Secrecy`].
+    ///
+    /// The secrecy of the result will be the [`Secrecy`] of the inputs.
+    pub fn mul<S: Secrecy>(
+        a: &[Scalar<S, impl ZeroChoice>],
+        b: &[Scalar<S, impl ZeroChoice>],
+    ) -> Vec<Scalar<S, Zero>> {
+        let mut result = vec![Scalar::zero(); a.len() + b.len() - 1];
+        for (i, &coeff_a) in a.iter().enumerate() {
+            for (j, &coeff_b) in b.iter().enumerate() {
+                result[i + j] = s!(result[i + j] + coeff_a.mark_zero() * coeff_b.mark_zero())
+                    .set_secrecy::<S>();
+            }
+        }
+
+        result
+    }
+
+    /// Adds two scalar polynomials together where `apoly` is mutated to form the result.
+    pub fn add_in_place<SA: Secrecy>(
+        apoly: &mut Vec<Scalar<SA, Zero>>,
+        bpoly: impl IntoIterator<Item = Scalar<impl Secrecy, impl ZeroChoice>>,
+    ) {
+        let b = bpoly.into_iter();
+        for (i, b) in b.enumerate() {
+            if i == apoly.len() {
```

Comment:
at first glance i thought this was wrong because of the `==` for unequal length polys, but the following push makes it repeat this branch on the next iteration :+1: